### DISCUSSION TITLE
Numpyro support?

### DISCUSSION THREAD

**Discussion by Chengwei94** [thomaspinder/GPJax#332] [General]
Is Gpjax compatible with numpyro? I have notice that there were once notebooks for numpyro and gpjax but it is no longer there. Does this means that gpjax no longer can be used with numpyro?

**Comment by daniel-dodd** [comment]
Thanks for the question @Chengwei94. 

The latest GPJax releases `v0.6.x` don't explicitly support NumPyro in either our code or documentation. However, given we use TensorFlow probability distributions, it may be possible to wrap these (as done here https://num.pyro.ai/en/0.7.0/_modules/numpyro/contrib/tfp/distributions.html) for NumPyro compatibility. 

We are keen to see a NumPyro integration though in the near future - this is on our radar. 

It would be nice to e.g.,

-  move all distributions from TensorFlow probability to NumPyro, or...
- ... provide a method to return NumPyro distribution's on objects instead of TensorFlow probability.

The first of these is something we tried back last Autumn #121, but there were a few issues and dropped it to push forward with releasing new features. With the new `gpjax.base.Module` introduced in `v0.6`, this should be easier to do this time round.

**Reply by iammteeh** [reply]
Hey Daniel,

just wanted to ask if there's any update on this?

kind regards

**Reply by thomaspinder** [reply]
Hey @iammteeh, we’re currently cleaning up some dependencies in the package. Once this is done, I’ll be looking into the numpyro integration. Is this something you’d be interested in collaborating with?

**Reply by iammteeh** [reply]
Hey @thomaspinder,

yes, would be nice if I can help! I developed a (fairly simple) GP using JAX and Numpyro in my master's thesis, before I discovered your project and it would be great to adapt to it while still be able to use MCMC samplers from Numpyro :)

Would be nice to know in which issues you ran!

**Reply by juanitorduz** [reply]
I am also interested in the integration, so I will be happy to help or test :) (I wrote most of https://num.pyro.ai/en/stable/contrib.html#hilbert-space-gaussian-processes-approximation )

**Reply by thomaspinder** [reply]
Awesome! The dependency issues are now resolved, and I have made a start @juanitorduz / @iammteeh on the numpyro migration. The backend (bijectors and distributions) of GPJax are now numpyro objects. The outstanding thing to to do is 1) work our how priors may be assigned to parameters, and 2) if it is possible right now to sample from GPJax's posterior using the current setup. From reading [NNX and Numpyro integration](https://num.pyro.ai/en/latest/tutorials/nnx_example.html), I think that it should be. 

If you have any thoughts, or would like to assist in either 1 or 2, then please message here. For 1), we'll need to modify the base `Parameter` class, whilst 2) may be resolved with a notebook demonstrating the integration.

**Reply by thomaspinder** [reply]
Sharing notes here:
- Numpyro's `nnx_module` cannot be directly used at it splits on `nnx.Param` whereas we subclass the parent class `nnx.Variable`. [Line 472](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/contrib/module.py#L472).
    - Switching `Param` for `Variable` appears to resolve the issue. 
- I have begun a notebook for [GPJax - Numpyro integration here](https://github.com/JaxGaussianProcesses/GPJax/blob/numpyro-integration-notebook/examples/numpyro_intergation.py). Right now, the issue appears to be that observation noise is not being correctly sampled (see image below). 
![image](https://github.com/user-attachments/assets/1c5c5722-80f1-4e9c-971a-ddbb2ae1fed9)

**Reply by juanitorduz** [reply]
Hi @thomaspinder ! Nice work!  Let me look into the notebook example this week :) I will be happy to help / review either (1) or(2). Let me first get familiarized with the integration strategy and take it from there :)

**Reply by thomaspinder** [reply]
Sounds great @juanitorduz. If you take a critical look at my first attempt, then please feel free to open a PR there with improvements or fixes, should you seen anything.

**Reply by juanitorduz** [reply]
One thing I see is that if we set (`smooth=False`) in the `plot_hdi` function, we see just get the training data 

![image](https://github.com/user-attachments/assets/91cf7c45-54a2-4a81-bda4-b09c4ff54797)

I also suggesting the `obs` parameter in the model

```python
    numpyro.sample(
        "likelihood",
        dist.MultivariateNormal(loc=prior_mean.squeeze(), covariance_matrix=prior_cov),
        # obs=data.y.squeeze(),    <- THIS!
    )
```

to try to fix this ...

When I run the notebook with these changes, here are the prior and posterior predictive

![image](https://github.com/user-attachments/assets/2e0eb520-cd0e-4f20-8e3c-a867bae03942)

![image](https://github.com/user-attachments/assets/64681d14-d7c8-4a8d-abbb-3cf3cf069a96)

Maybe the priors are too conservative ... 🤔

**Reply by juanitorduz** [reply]
If I remove the `obs` fromt the distribution as above, I think I need to condition in the squeezed vector `conditioned_model = condition(model, data={"likelihood": y_train.squeeze()})`. Still, for the posterior predictive, I keep getting the flat lines

![image](https://github.com/user-attachments/assets/8337778a-cbd3-4e3a-bf8d-fc6be687e677)

Even though the parameters seem ok 

![image](https://github.com/user-attachments/assets/48823214-efd6-4ed4-8a9a-8221a5ef2095)

(We need to use the "gp/" prefix in the `return_sites` parameter)

I wonder if we need additional steps to generate posterior predictive samples from the likelihood  .... In https://juanitorduz.github.io/hsgp_intro/ I did a GP from scratch and I need to use the condition formula to generate posterior predictive samples. I remember I tried to do it with a Multivariate Normal and I had similar results (flat lines). Is this something that GPJax does under the hood? 🤔

**Reply by thomaspinder** [reply]
Yes - we would do the conditioning [here](https://github.com/JaxGaussianProcesses/GPJax/blob/main/gpjax/gps.py#L444).

**Reply by thomaspinder** [reply]
I had also wondered if we should be constraining/unconstraining the module.

**Reply by juanitorduz** [reply]
So we need a to use the "predict" method somewhere in the model right?

**Reply by thomaspinder** [reply]
Yes - I suspect it is similar to the prediction we do [here](https://docs.jaxgaussianprocesses.com/_examples/poisson/#prediction) when integrating with Blackjax. 

Feel free to share code if you'd like me to simultaneously investigate.

**Reply by juanitorduz** [reply]
Here is my working branch https://github.com/juanitorduz/GPJax/blob/gpjax_numpyro_test/examples/numpyro_intergation.py (no major changes). I wonder if we could do the `predict` inside the numpyro model definition.

**Reply by thomaspinder** [reply]
> I wonder if we could do the predict inside the numpyro model definition.

Excuse my numpyro ignorance, but what would the benefit of this be?

**Reply by juanitorduz** [reply]
Welll, at least for me, I think it would be great to be able to have a GP component as a latent variable as in https://juanitorduz.github.io/electricity_forecast/ . But maybe I am missing something 🙈 . 

In PyMC there is the `gp.prior` and `gp.conditional` to do this https://www.pymc.io/projects/examples/en/latest/gaussian_processes/GP-Latent.html so its a two-step approach. Maybe it will work the same with GPJax (following the BlackJAX example)?

**Reply by juanitorduz** [reply]
(maybe I was not meaning predict, but rather the condition step)

**Reply by thomaspinder** [reply]
Ah, gotcha. Yes, that is also the end state that I would like this integration to be in! For now, I’d be very happy if we can simply get a reasonable evaluation of the predictive posterior

**Reply by thomaspinder** [reply]
A worked example like your electricity forecasting example though we be a wonderful end state of this integration. That post really gets at the functionality that integrating GPJax with numpyro could offer

**Reply by thomaspinder** [reply]
The challenge I am having right now is working out a clean way to map the samples generated by numpyro back into a Flax/NNX state. If this were possible, then the prediction step would work.

**Reply by juanitorduz** [reply]
I am also looking into this 😄. I'll need to sleep it over 😅. Keep me posted 💪

**Reply by thomaspinder** [reply]
Likewise, should you have any productive ideas here!

**Reply by thomaspinder** [reply]
https://github.com/google/flax/discussions/4704#discussioncomment-12834070

**Reply by juanitorduz** [reply]
@thomaspinder In https://github.com/juanitorduz/GPJax/commit/b42ed96e6b182b15ac7b7c9040ebe3304e8cde29 I have a brute force approach to parse the samples into the state template (I am pretty sure there should be a nicer version using `nnx.transversals` as suggested) (Update: https://github.com/juanitorduz/GPJax/commit/8e1209dcfb44a562faf8f0919d6df8cd9071a1a9#commitcomment-155494638)
